app/models/subscription/BusinessSubscriptionDetailsModel.scala

```diff
@@ -0,0 +1,198 @@
+                               ukProperty: Boolean,
+                               foreignProperty: Boolean)
+
+object FeIncomeSourceModel {
+  implicit val format: Format[FeIncomeSourceModel] = Json.format[FeIncomeSourceModel]
+}
+
+case class PropertyCommencementDateModel(startDate: DateModel)
+
+object PropertyCommencementDateModel {
+  implicit val format: OFormat[PropertyCommencementDateModel] = Json.format[PropertyCommencementDateModel]
+}
+
+case class AccountingMethodPropertyModel(propertyAccountingMethod: AccountingMethod)
+
+object AccountingMethodPropertyModel {
+  implicit val format: OFormat[AccountingMethodPropertyModel] = Json.format[AccountingMethodPropertyModel]
+}
+
+case class OverseasPropertyCommencementDateModel(startDate: DateModel)
+
+object OverseasPropertyCommencementDateModel {
+  implicit val format: OFormat[OverseasPropertyCommencementDateModel] = Json.format[OverseasPropertyCommencementDateModel]
+}
+
+case class OverseasAccountingMethodPropertyModel(overseasPropertyAccountingMethod: AccountingMethod)
+
+object OverseasAccountingMethodPropertyModel {
+  implicit val format: OFormat[OverseasAccountingMethodPropertyModel] = Json.format[OverseasAccountingMethodPropertyModel]
+}
+
+case class BusinessSubscriptionDetailsModel(accountingPeriod: AccountingPeriodModel,
+                                            selfEmploymentsData: Seq[SelfEmploymentData],
+                                            accountingMethod: AccountingMethod,
+                                            incomeSource: FeIncomeSourceModel,
+                                            propertyCommencementDate: Option[PropertyCommencementDateModel] = None,
+                                            propertyAccountingMethod: Option[AccountingMethodPropertyModel] = None,
+                                            overseasPropertyCommencementDate: Option[OverseasPropertyCommencementDateModel] = None,
+                                            overseasAccountingMethodProperty: Option[OverseasAccountingMethodPropertyModel] = None
+                                           )
+
+object BusinessSubscriptionDetailsModel {
+
+  implicit val businessSubscriptionReads: Reads[BusinessSubscriptionDetailsModel] = Json.reads[BusinessSubscriptionDetailsModel]
+
+  implicit val businessSubscriptionWrites: Writes[BusinessSubscriptionDetailsModel] = new Writes[BusinessSubscriptionDetailsModel] {
+    def writes(details: BusinessSubscriptionDetailsModel) = Json.obj(
+      "businessDetails" -> details.selfEmploymentsData.map(
+        data => Map("accountingPeriodStartDate" -> JsString(details.accountingPeriod.startDate.toDesDateFormat),
+          "accountingPeriodEndDate" -> JsString(details.accountingPeriod.endDate.toDesDateFormat),
+          "tradingName" -> JsString(data.businessTradeName.map(_.businessTradeName).get),
+          "addressDetails" -> Json.obj(
+            "addressLine1" -> data.businessAddress.map(_.address.lines.head).get,
```

Comment:
This would be better if we use .getOrElse(throw new Exception("Missing
 business address parameter")) as it would put an actual error in the logs if it ever happened somehow.

Reply:
Done